### PR TITLE
Feature/resource manager

### PR DESCRIPTION
**Pull request by Renardjojo**
Imported resource manager and wrapper of the resource manager. The scene manager is not done for now

### CODE REVIEW

**Review by Korotsu — APPROVED** [review]

File: engine/include/Engine/Resources/Scene.hpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+        //TODO: Scene can be copiable ? How manage resource
+        constexpr inline
+        Scene (const Scene& other) noexcept			    = delete;
+    
+        //TODO: Scene can be move ? How manage resource
+        constexpr inline
+        Scene (Scene&& other) noexcept				    = delete;
+        
+        //TODO: Scene can be capiable ? How manage resource
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/Scene.hpp]:
```suggestion
        //TODO: Can scene be copied ? How to manage resource
```

File: engine/include/Engine/Resources/Scene.hpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    
+        //TODO: Scene can be create by default ?
+        inline
+        ~Scene () noexcept				                = delete;
+
+        //TODO: Scene can be copiable ? How manage resource
+        constexpr inline
+        Scene (const Scene& other) noexcept			    = delete;
+    
+        //TODO: Scene can be move ? How manage resource
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/Scene.hpp]:
```suggestion
        //TODO: Can scene be moved ? How to manage resource
```

File: engine/include/Engine/Resources/Scene.hpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        Scene (ResourcesManager<int>& resourceManager) noexcept
+        {
+            loadResource(resourceManager);
+        }
+    
+        //TODO: Scene can be create by default ?
+        inline
+        ~Scene () noexcept				                = delete;
+
+        //TODO: Scene can be copiable ? How manage resource
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/Scene.hpp]:
```suggestion
        //TODO: Can scene be copied ? How to manage resource
```

File: engine/include/Engine/Resources/Scene.hpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    
+        public:
+    
+        constexpr inline
+        Scene (ResourcesManager<int>& resourceManager) noexcept
+        {
+            loadResource(resourceManager);
+        }
+    
+        //TODO: Scene can be create by default ?
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/Scene.hpp]:
```suggestion
        //TODO: Can scene be created by default ?
```

File: engine/include/Engine/Resources/Scene.hpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    
+        //TODO: Scene can be move ? How manage resource
+        constexpr inline
+        Scene (Scene&& other) noexcept				    = delete;
+        
+        //TODO: Scene can be capiable ? How manage resource
+        constexpr inline
+        Scene& operator=(Scene const& other) noexcept		= delete;
+    
+        //TODO: Scene can be move ? How manage resource
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/Scene.hpp]:
```suggestion
        //TODO: Can scene be moved ? How to manage resource
```

File: engine/include/Engine/Resources/SceneManager.hpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    enum class ESceneGraphManagement
+    {
+        REPLACE,            //Change the scene graph
+        MERGE,              //Merge both scene graph
+    };
+
+    enum class EResourceManagement
+    {
+        RECYCLING,          //Enable recycling process
+        KEEP_IN_MEMORY,     //Keep ALL resources in memory even if there are not used. Usefull to load fastely previous scene
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/SceneManager.hpp]:
```suggestion
        KEEP_IN_MEMORY,     //Keep ALL resources in memory even if they are not used. Usefull to load fastly previous scene
```

File: engine/include/Engine/Resources/ResourcesManager.hpp [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * @brief add element with argument of constructor in parameter and key. Resource canno't have the same key
+     *
+     * @tparam Args
+     * @param args
+     * @return unsigned int
+     */
+    template <typename... Args>
+    LType& add(std::string key, Args&&... args) noexcept;
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/ResourcesManager.hpp]: What if the resource constructor has an exception ?
invalid noexcept

File: engine/include/Engine/Resources/ResourcesManager.hpp [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * @brief Remove resource with the corresponding key if key is found
+     * @param key 
+     * @return 
+    */
+    inline void remove(const std::string& key) noexcept;
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/ResourcesManager.hpp]: What if the resource destructor has an exception ?
invalid noexcept

File: engine/include/Engine/Resources/ResourcesManager.hpp [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * @brief clear the container and all associate resources
+     * @return 
+    */
+    inline void clear() noexcept;
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/ResourcesManager.hpp]: What if the resource destructor has an exception ?
invalid noexcept

File: engine/include/Engine/Resources/ResourcesManager.hpp [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * @brief Remove resource with the corresponding key if key is found
+     * @example remove<Texture>("AwsomeTexture")
+     * @tparam T 
+     * @param key 
+     * @return 
+    */
+    template <class T>
+    inline void remove(const std::string& key) noexcept;
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/ResourcesManager.hpp]: What if the resource destructor has an exception ?
invalid noexcept

File: engine/include/Engine/Resources/ResourcesManager.hpp [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    /**
+     * @brief Add resource.
+     * @example add<Texture>("AwsomeTexture", Texture())
+     * @tparam T
+     * @tparam Args
+     * @param args
+     * @return unsigned int
+     */
+    template <class T, typename... Args>
+    inline T& add(const std::string& key, Args&&... args) noexcept;
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/ResourcesManager.hpp]: What if the resource constructor has an exception ?
invalid noexcept

File: engine/include/Engine/Resources/ResourcesManager.hpp [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * @brief clear the container and all associate resources
+     * @example clear<Texture>()
+     * @tparam T 
+     * @return 
+    */
+    template <class T>
+    inline void clear() noexcept;
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/ResourcesManager.hpp]: What if the resource destructor has an exception ?
invalid noexcept

File: engine/include/Engine/Resources/Scene.hpp [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        friend class SceneManager;
+
+        private:
+    
+        protected:
+
+        ResourcesManager<int> m_resourceManager;
+
+        constexpr
+        void loadResource(ResourcesManager<int>& resourceManager) noexcept {};
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/Scene.hpp:24]: What if the resource constructor has an exception ?
invalid noexcept

File: engine/include/Engine/Resources/Scene.hpp [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    
+        public:
+    
+        constexpr inline
+        Scene (ResourcesManager<int>& resourceManager) noexcept
```
Comment (ThomasProg) [inline engine/include/Engine/Resources/Scene.hpp:29]: What if one of the resource constructor has an exception ?
invalid noexcept

File: engine/include/Engine/Resources/Scene.hpp (hunk 4 above)
Comment (ThomasProg) [inline engine/include/Engine/Resources/Scene.hpp]:
```suggestion
        //TODO: Can scene be created by default ?
```

**Review by ThomasProg — APPROVED** [review]
noexcept should be reviewed, but it is not a big deal for this project so the pull request is accepted.
It is not said functions should not be noexcept, but they might need conditional noexcept depending on constructors, destructors, etc.

**Pull request merged**